File: js/libraries/shared/src/common/path.ts
// NODE ONLY!

import path from 'path';
import os from 'os';
import fs from 'fs';

export function lookupAssetExchangePath(createDirMode: 'all' | 'parent' | 'none', ...paths: string[]): string {
  // construct path
  const result = path.join(os.homedir(), '.assetexchange', ...paths);

  // create directories on request
  switch (createDirMode) {
    case 'all':
      fs.mkdirSync(result, { recursive: true });
      break;
    case 'parent':
      fs.mkdirSync(path.dirname(result), { recursive: true });
      break;
  }

  return result;
}

export function lookupServicesPath(createDirMode: 'all' | 'parent' | 'none', ...parts: string[]): string {
  return lookupAssetExchangePath(createDirMode, 'services', ...parts);
}

export function lookupServiceEntryPath(category: string, type: string, pid: number): string {
  return path.join(lookupServicesPath('all', category, type), pid + '');
}

File: go/libs/shared/common/path.go
package common

import (
	"os"
	"path"
	"strconv"
)

type CreateDirMode int

const (
	CreateDirModeAll = iota
	CreateDirModeParent
	CreateDirModeNone
)

func LookupAssetExchangePath(createDirMode CreateDirMode, paths ...string) (string, error) {

	// construct path
	homePath, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}

	result := path.Join(append([]string{homePath, ".assetexchange"}, paths...)...)

	// create directories on request
	if createDirMode == CreateDirModeAll {
		err = os.MkdirAll(result, 0755)
		if err != nil {
			return "", err
		}
	} else if createDirMode == CreateDirModeParent {
		err = os.MkdirAll(path.Dir(result), 0755)
		if err != nil {
			return "", err
		}
	}

	return result, nil

}

// LookupServicesPath creates a path within the rpc registry and returns the normalized path
func LookupServicesPath(createDirMode CreateDirMode, parts ...string) (string, error) {
	return LookupAssetExchangePath(createDirMode, append([]string{"services"}, parts...)...)
}

// LookupServiceEntryPath returns the path to a registration file within the rpc registry
func LookupServiceEntryPath(category string, typex string, pid int) (string, error) {
	dir, err := LookupServicesPath(CreateDirModeAll, category, typex)
	if err != nil {
		return "", err
	}
	return path.Join(dir, strconv.Itoa(pid)), nil
}
